### ISSUE TITLE
SFT with answer_only_loss: auto-slice output vocab for faster training

### ISSUE THREAD

**Issue by AmitMY** [NVIDIA-NeMo/Megatron-Bridge#2473]
## Problem

When fine-tuning with `answer_only_loss`, the output projection (`logits = hidden @ embedding.T`) and cross-entropy loss are computed over the **full vocabulary**, even though the model may only need to predict tokens from a small subset of the vocab. This is wasteful — the output layer matmul and softmax scale linearly with vocab size.

For example:
- **Qwen3-0.6B** has 151,936 tokens. If you fine-tune it for any-language → Hebrew translation, and Hebrew uses ~50 unique token IDs in answers, you're still multiplying hidden states by a 151,936-row matrix every step.
- **Our use case**: We combine a spoken-language byte-level tokenizer (260 tokens) with a video tokenizer (64,000 tokens) into a single 64,260-token vocabulary. During SFT (video → text), all answer tokens are byte-only (IDs 0–259). The input contains video tokens (IDs 260–64,259), so we can't resize the embedding matrix — but the output head only ever needs 260 of the 64,260 rows.

## Measured speedup

We monkey-patched the output layer during SFT to slice the shared embedding weight:

```python
def _byte_output_forward(input_, weight=None, **kwargs):
    weight = weight[:260]  # slice to answer-only vocab
    return F.linear(input_, weight), None

model.output_layer.forward = _byte_output_forward

# Also clamp labels at masked positions (video token IDs > 259
# would be out-of-bounds for the 260-dim logits, but loss_mask=0
# for those positions anyway):
original_loss = model.compute_language_model_loss
model.compute_language_model_loss = lambda labels, logits: original_loss(
    labels.clamp(max=259), logits
)
```

**Results** (60M model, SEQ=16384, packed sequences, single GPU):

| Config | Step time | TFLOP/s/GPU | Speedup |
|--------|-----------|-------------|---------|
| Full vocab (64,260) | 5.41s | 31.6 | 1.0x |
| Byte-only head (260) | 3.03s | 55.8 | **1.79x** |

Isolated benchmarks on the output layer alone show **210x** speedup on the matmul and **286x** on cross-entropy.

Even though we do not update the other embeddings at the LM head - Quality is preserved (chrF 0.0527 vs 0.0497 baseline at 1000 steps). The checkpoint is saved with the full vocab and converts to HF format normally — the slicing is training-only.

## Proposed generic solution

When `answer_only_loss` is enabled, Megatron-Bridge could automatically detect which token IDs actually appear in answers and slice the output projection accordingly:

1. **During data loading**: scan all answer spans (where `loss_mask=1`) and collect the set of unique token IDs → `active_ids` (a sorted tensor of, say, N unique IDs out of V total vocab).
2. **Patch output layer**: instead of `F.linear(hidden, weight)` with weight `[V, H]`, compute `F.linear(hidden, weight[active_ids])` with weight `[N, H]`.
3. **Remap labels**: map original label IDs to their index in `active_ids` (e.g., via a lookup table). Map all masked-position labels to 0 (or any valid index, since `loss_mask=0` anyway).

This gives `[batch, seq, N]` logits instead of `[batch, seq, V]`, and cross-entropy over N classes instead of V.

### Trade-off

This is not exactly equivalent to full fine-tuning because gradients only flow through the `active_ids` rows of the embedding matrix (for the output projection path). The input embedding still uses all V rows. In practice, for SFT where answers are a small subset of the vocab, this is unlikely to matter — and users who care can disable the optimization.

### Where this helps most

- Large vocab + small answer vocab (Qwen 151k → Hebrew 50 tokens)
- Multimodal models where input modality tokens dominate vocab but answers are text-only
- Any SFT task where the answer vocabulary is much smaller than the full tokenizer vocabulary

A flag like `auto_slice_output_vocab: true` (or even on-by-default for `answer_only_loss`) would make this transparent to users.

**Comment by yaoyu-33** [comment]
@AmitMY : thanks for the proposal. It looks pretty nice!
1. Do you have loss curve to share? auto slice on vs. off.
2. Are you interested in a contribution to repo directly? otherwise, we can assign someone to follow up on this.

**Comment by AmitMY** [comment]
i don't have any solid curve to share at the moment, especially since lots of iterations were made and i can't find the exact point of the switch.
Nevertheless, this is straightforward - and should actually start from a lower loss, because we are masking all the other tokens.

Claude made a PR #2522 but it can't run the actions due to
> AADSTS90072: User account 'amitmoryossef@gmail.com' from identity provider 'live.com' does not exist in tenant 'NVIDIA Corporation' and cannot access the application 'baaa4ff1-f56d-4f79-8a75-4c0cdf02e84a'(Starfleet Identity Federation (production, protected tier)) in that tenant. The account needs to be added as an external user in the tenant first. Sign out and sign in again with a different Azure Active Directory user account

**Comment by yaoyu-33** [comment]
sounds good. lemme check the pr.

**Comment by cuichenx** [comment]
@AmitMY  I ran benchmarking on real models and found out the speed up isn't significant when the majority of the flops is in the base transformer model, instead of the output projection layer.

https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3679#issuecomment-4392759648

Can you check if I missed anything?

**Comment by AmitMY** [comment]
@cuichenx This is a speedup only for a single layer.
If the transformer has more layers, you will see less of a huge speedup, as expected.
But, for small LMs (in my example, 60m parameters), the speedup is real (1.79x) because actually most of the compute is on the output layer.

This is really great for small scale experimentation.

#### What I think this could bring in the future to large models? 
Labs do not increase their embedding table sizes, not because it is hard to embed, but because it is hard to softmax. This results in Hebrew vocab, for example, being like 20/100,000 tokens in total. 
If a pathway for 1M tokens would be feasible, maybe we will see models which are more efficient for other languages.

**Comment by cuichenx** [comment]
The optimization makes sense for the specific setup you described, but we do not plan to add this to Megatron-Bridge.

Megatron-LM/Megatron-Bridge is primarily intended for large-scale training workloads, and we generally do not optimize the stack for models below roughly 500M parameters. For smaller models and this kind of task-specific training optimization, NeMo AutoModel may be a better fit and the right place to explore support.

Thanks again for the detailed proposal and measurements. Closing this as out of scope for Megatron-Bridge.
